### ISSUE TITLE
Bring back breakpad on osx

### ISSUE THREAD

**Issue by monkeyiq** [fontforge/fontforge#2120]
The build 201412/14_0426 marked the merge of breakpad for the client side on osx. Unfortunately the breakpad functionality stopped working on 201412/20_0427 and later builds. I notice a commit at around that time https://github.com/fontforge/fontforge/commit/07834e53d33dd5356b468569edb4873cd700e1ec which disables listen_to_apple_events. Disabling apple_events also disables a whole block which implements the main loop on osx, falling back to the general main loop: [startui.c#L1359](https://github.com/fontforge/fontforge/blob/master/fontforgeexe/startui.c#L1359).

I'm happy to not rely on RunApplicationEventLoop() on osx, but instead, after installing the apple_event handlers and the cocoa application, just use the same event handling main loop as Linux and Win use. Unfortunately, if you disable the cocoa application setup then you also disable the breakpad handling which wants to hook into the objective-c Application object [MacBreakpadStarterGuide](http://code.google.com/p/google-breakpad/wiki/MacBreakpadStarterGuide#Write_Code!).

@frank-trampe I realize that it's months on now, but do you recall which issues were being solved by https://github.com/fontforge/fontforge/commit/07834e53d33dd5356b468569edb4873cd700e1ec so that I can try to keep an eye out for them when I reenable part of the apple_events code.

**Comment by adrientetar** [comment]
> @frank-trampe I realize that it's months on now, but do you recall which issues were being solved by 07834e5 so that I can try to keep an eye out for them when I reenable part of the apple_events code.

See #1956, it was to stop issues presumably caused by deprecated Carbon APIs on the Mac. It seems using Cocoa instead would fix this.

**Comment by adrientetar** [comment]
PR #2004, and #2058 is a followup to that issue.

**Comment by monkeyiq** [comment]
@adrientetar Thanks for the clarification! I may then also make the change reenabling breakpad also remove the RunApplicationEventLoop() declaration and usage. There seems little point leaving that declaration of an osx function around if it's usage should now be avoided.

**Comment by frank-trampe** [comment]
@monkeyiq, the problem was that users of newer Macintosh versions were having trouble with the application pausing frequently when the event loop was enabled.

**Comment by frank-trampe** [comment]
@monkeyiq, I would prefer to keep the disabled code in place until we replace it in case it is helpful in formulating said replacement.

**Comment by monkeyiq** [comment]
@frank-trampe I think I'll add some comment love there then. It seems that if RunApplicationEventLoop() is doing these undesirable pauses then it should be readily stated in the startui.c file just in case somebody else is looking at this in a few years time and gets tempted to re-enable that function call. It's possible that reenabling the call at the time might even work ok for 6 months until the other side of the function changes again in 10.13 or whatever.

It seems to be running ok with all the other code than RunApplicationEventLoop() reenabled.

**Comment by monkeyiq** [comment]
ok, I merged the above PR. If the pausing happens then I'll take a knife to it and trim back some more to be closer to listen_to_apple_events=false;

**Comment by monkeyiq** [comment]
With merge this is uploading to 201503/07_1237
If anyone finds slow downs / pauses again with this please respond here and I'll trim a bit more of the osx specific stuff again.

**Comment by monkeyiq** [comment]
Breakpad is now back and running on the osx builds. To test one can generate a crash by "File/Execute Script" and select "FF" style with the script simply as

DebugCrashFontForge()

This will drop all the GUI very quickly with a dialog offering to upload the crash report appearing after a short delay. Clicking on "Send Report" leaves the window open for a moment and then it closes.

on the bigv machine, under the ffbugreports user you should then notice a new report in the directory 
/home/ffbugreports/report-acceptor/uploads

each new report generates three files. The first is the minidump from the client. The second is some metadata from the client, for example the git hash of the build they were using at the time the minidump was generated. The server side script also unpacks the "dmp" file into a backtrace using the symbol files that are extracted at osx bundle build time. If the backtrace finds the right symbols then file names and line numbers become available in the backtrace file. Quite a bit cooler for FontForge users to be able to tell devs where the crash was :)

e753ff92cdbe03fbad6e9f2202a316a1.dmp
e753ff92cdbe03fbad6e9f2202a316a1.dmp.json
e753ff92cdbe03fbad6e9f2202a316a1.dmp.backtrace

I have tested this with the 201503/09_2117 build.

The script osx/create-osx-app-bundle.sh does the breakpad symbol file generation and renames things into the expected tree format. So if folks want to do private or local only builds, they can also still separate out the symbol files, though they will also want to tweak the breakpad config in FontForge to upload the minidump to a local server so they can then bring that all together to make the backtrace. Though for local builds using lldb with full debug might be a simpler process.

Anyway, the next plans are to upload the backtrace + json files from bigv to somewhere to allow tracking. I'm thinking maybe to github and trigger a new issue to be made with the backtrace. One problem with github is which user account to use to do the it push and other automation. It seems github new user policy goes against allowing an account for an upload bot. IMHO the gh account used should be an account with only access to a ffbugreports repo on github to limit the potential of hacks. I might even email them to see if its cool to make a ffbugreports account as a free account even though I already have one.

I have tried to enable RequestComments for breakpad to allow the user to type in what they were up to before the crash. Unfortunately, while enabling a request for comment or email creates a suitable dialog, that dialog doesn't capture keyboard events. I've tried a few things to try to get around that, but haven't had success yet. It would be handy on a few levels to allow a user comment, one being the ability for the user to then find the backtrace on github and monitor it or add more information. The main github doesn't have the comments enabled to avoid infuriating the user with a non typable comment box.

Of coarse the third task is getting Linux and Win to do the breakpad client dance too :)

**Comment by davelab6** [comment]
Why not remove bigv from the process and bundle the minidump processor and the symbols file and whatever else in the app bundle, and then initially write the files to a configurable dir (default `~`) and then use the person's own github account to post the line numbered files to github issue (and/or gist) via https://developer.github.com/v3/

**Comment by davelab6** [comment]
That way the backtrace is guaranteed to find the right symbols, and the file named and line numbered backtrace is the best and only result that we need from breakpad - right?

**Comment by adrientetar** [comment]
> It seems github new user policy goes against allowing an account for an upload bot.

There are numerous bots on GitHub, e.g. @bors.

**Comment by monkeyiq** [comment]
Doing things locally on the user machine might work out simpler... I'll have a look at how things are implemented in bp. IIRC platforms have unique client side in bp which are all aimed one way or another at sending the minidump to a http(s) port on the web.

For osx the breakpad symbols are also shipped, and are at a known location. So a patch to the crash reporter app for osx in bp will be able to find them, process them, and then maybe open a finder window showing them or do some gh api stuff to directly create the new issue. One thing that sticks out about the final idea is the whole web api authentication, "fontforge crash reporter wants access to your github account" if using the API directly from the crash reporter.

**Comment by davelab6** [comment]
What sticks out? Seems normal to ask for a login to post data :) github has
ssh keys to authenticaye too, sparkleshare does it that way. Oauth might
also be possible?

**Comment by davelab6** [comment]
Right, but standing up a httpd with python httpd in std lib is easy to take
the data file without altering bp

**Comment by monkeyiq** [comment]
using python and a local http is an interesting idea. A think out loud of the lifecycle for handing this... At the point that the breakpad osx crash reporter is running FontForge has obviously already crashed.

Given that FontForge is started from a script, that  script is free to start the localhost python bp endpoint before running FontForge itself. If FontForge closes regularly then we can kill the python endpoint from the script. If fontforge crashes then we can force breakpad to submit the report (thus it will make no "send report" dialog itself). The python http server will then get called with the report, and we can assume that it will massage the report and send it on through before quitting. The main thing is that the launch script must detect a segv result of running FontForge and *not* kill the python http endpoint process in that case. 

I'll move to get that going first, then improve the python httpd script.

**Comment by davelab6** [comment]
On 12 March 2015 at 05:56, monkeyiq <notifications@github.com> wrote:

> The main thing is that the launch script must detect a segv result of
> running FontForge and *not* kill the python http endpoint process in that
> case.
>
> I'll move to get that going first, then improve the python httpd script.
>
If the launch script moves from shell to python, that would be okay.

However, looks like the bash trap command could be used for this.

http://unix.stackexchange.com/questions/24307/how-can-i-trap-a-program-that-returns-139-segmentation-fault-in-bash

Or,

```
some-crashy-program argsif [ $? -eq 139 ]; then
    echo "It crashed!"
    exit 1fi
```

**Comment by monkeyiq** [comment]
Unfortunately with breakpad you get a $? of 2 on crash. For a trivial example it does come back as 128+segv.

**Comment by davelab6** [comment]
Well the int to compare is different but will the technique work?

**Comment by monkeyiq** [comment]
interestingly you can redirect a browser to a url to create a new issue on gh. The unfortunate part is that trying to marshal newlines through that hoop is not trivial. And there are many levels where the newline could be being eaten at. I can't find an explicit markdown <br/> equiv either, so cheating seems hard.

**Comment by monkeyiq** [comment]
well, one idea is to have the python handler cleaned up right away on return code of 0. For 2 we don't know if its breakpad or something else that caused ff to exit with 2. So it seems that degenerating to a 5 minute timeout might be the way to go there. The breakpad http server script should be done within 5 or 15 or 30 minutes, so then we can be sure to clean it up. Unless the user restarts ff (quite likely) in which case we can't spawn two python http server crash handlers because they want the same port, so it's another case.

**Comment by monkeyiq** [comment]
The big gain of redirecting the browser to the gh/ff/issues/new URL is that no OAuth is needed, and no special tokens then have to be protected. The browser handles that part, user logs in and report is filled.

Downside;
* can only do about 4-6kb of backtrace because of GET size limits
* newlines keep getting eaten for me right now.

The top 4kb gives the bottom part of the backtrace and reasonable scope, so that might even be ok. The newlines being stripped makes the default issue very hard to read as a human though.

It was a good pipedream for UX for ff users IMHO -- Not needing to do OAuth just for a new bug report. But OAuth might end up being the cleaner way to submit. With the caveat of how to protect the oauth token.

**Comment by monkeyiq** [comment]
actually, I think I have the core of this figured.

**Comment by monkeyiq** [comment]
An example with a fairly ugly intermediate web page is uploading to 201503/13_2155

See osx/FontForge.app/Contents/Resources/breakpad/webinterface for the current location of the web interface files. The index.html uses handlebar like substitution and the script will allow you to load png, css, and js files with a very limited naming scheme from the webinterface directory. The file name limitations are in place to avoid serving files that are not in that directory.

files like foo.png and alice.css are fine, and such simple file names should be ok if they are just used to build this intermediate web page.

The github link should take you there, possibly needing you to login as normal, and a new issue will be setup with a body showing the backtrace.

So, in other words this is now getting close to being mergable. The index.html obviously could use liberal sprinklings of love to look better, but others are likely to know how to do that better than me. If we want to merge a template in then I can help lift or modify the file name restrictions for that part.

I might also push gist link support for folks who don't want to log in to anything.

**Comment by davelab6** [comment]
> The file name limitations are in place to avoid serving files that are not in that directory.

where are they set, exactly?

> I might also push gist link support for folks who don't want to log in to anything.

That would be great - because...

> The index.html uses handlebar like substitution 

...what is the data size limitation on this? I guess that the backtrace should be posted on gist always, and then the issue should be created linking to that

**Comment by davelab6** [comment]
Nice! Just tested this, works great! #2141 was the result of running the FF script `DebugCrashFontForge()` :) 

- [ ] Set title to "FontForge Breakpad Crash Report"
- [ ] Set the first line of the report to `< Please replace this line with a description of what actions you took just before the crash happened - the technical details below will help the fontforge developer community to fix the problem, but knowing what happened just before the crash is also helpful >` then 5 blank lines; the `< >` brackets mean that the text won't be visible if they forget to write it

**Comment by monkeyiq** [comment]
201503/14_1215 now with a gist upload option. seems that I can't search for the uploaded gists though, even though I mark them public :/

**Comment by monkeyiq** [comment]
I'll take a look at this slow launch thing tonight. I haven't noticed anything myself so far :| Does the terminal say anything while you are waiting if you run it there?

**Comment by monkeyiq** [comment]
file name limitations are set at:
https://github.com/fontforge/fontforge/pull/2140/files#diff-00c1d4ba207846bcf10c12d426f4b343R71

they are harsh because its much easier to be secure with a block by default than to try to quickly think of all the little ways somebody might try to do a /foo/../../../secret.txt type attack.

**Comment by monkeyiq** [comment]
For the above, the handlebar substitutions are now expanded. The default is about the first 4kb of the backtrace which is also touched up for direct github issue creation.

the new report_raw is the entire backtrace with the json metadata at the top, unmangled.
https://github.com/fontforge/fontforge/pull/2140/files#diff-00c1d4ba207846bcf10c12d426f4b343R105
the raw version is what is uploaded to gist.

The gist upload already redirects to the uploaded gist page for you
https://github.com/fontforge/fontforge/pull/2140/files#diff-42d7b761bcb12abb5323e06b63af70d8R35

so the github version should post to gist, then instead of redirecting to the html gist page, embed that url into the new github report. It is likely handy to inject the user feedback into the gist and the github issue so its easy to find in both.

**Comment by monkeyiq** [comment]
For the "Set the first line of the report to..." I went a slightly different way. jQuery and bootstrap are now available to the crash report web page which is now updated a bit to use that. The gain of getting the description there is that it can also be posted to gist or whereever.

I've also updated the github part to first post the full backtrace to gist and then include a link to that gist instead of the whole backtrace in the github issue. I'll leave the line commented there in case we want the first 4kb back on the github issue as well.

**Comment by davelab6** [comment]
Please pop the latest build on fuuko and ill check it out :) now for win
then Ubuntu

**Comment by monkeyiq** [comment]
cool, just pushing it up to gh now for the next build ;)

**Comment by monkeyiq** [comment]
this will be 201503/14_2127... uploading

**Comment by monkeyiq** [comment]
hmm, doesn't seem to find bootstrap in that build :| grr.

**Comment by monkeyiq** [comment]
grr, simple prob, found fix

**Comment by monkeyiq** [comment]
ok ... uploading ... 201503/14_2204

**Comment by monkeyiq** [comment]
merged to mainline. a new master build soon.
